client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -149,17 +205,25 @@ private final class PoolManager[F[_], A <: Connection[F]](
 
             case None if waitQueue.isEmpty =>
               logger.debug(s"Returning idle connection to pool: $stats")
-              idleQueue.enqueue(connection)
+              addToIdleQueue(connection, key)
 
-            // returned connection didn't match any pending request: kill it and start a new one for a queued request
             case None =>
-              connection.shutdown()
-              allocated -= 1
-              val Waiting(key, callback) = waitQueue.dequeue()
-              createConnection(key, callback)
+              waitQueue.dequeueFirst { waiter =>
+                allocated.getOrElse(waiter.key, 0) < maxConnectionsPerRequestKey(waiter.key)
+              } match {
```

Comment:
I think this is right now for this case.  I think this function can be extracted and reused for a `dequeueFirst` on line 236.

I think down there, the Some case just creates a connection, and the None case is a no-op.  So maybe it's as simple as:

```
waitQueue.dequeueFirst { f => _}.map { case Waiting(k, cb) => createConnection(k, cb) }
```

I don't understand @bryce-anderson's [comment](https://github.com/http4s/http4s/pull/1367/files#r139319549) about a different predicate.  I think the logic he's talking about on not closing the connection is handled here, and the connection is already closed down there.